sql/telemetry_derived/clients_histogram_bucket_counts_v1/query.sql

```diff
@@ -40,61 +41,166 @@ RETURNS ARRAY<STRUCT<key STRING, value FLOAT64>> AS (
   )
 );
 
-CREATE TEMP FUNCTION udf_normalize_histograms (
-  arrs ARRAY<STRUCT<
-    first_bucket INT64,
-    last_bucket INT64,
-    num_buckets INT64,
-    latest_version INT64,
-    metric STRING,
-    metric_type STRING,
-    key STRING,
-    process STRING,
-    agg_type STRING,
-    aggregates ARRAY<STRUCT<key STRING, value INT64>>>>)
-RETURNS ARRAY<STRUCT<
-  first_bucket INT64,
-  last_bucket INT64,
-  num_buckets INT64,
-  latest_version INT64,
-  metric STRING,
-  metric_type STRING,
-  key STRING,
-  process STRING,
-  agg_type STRING,
-  aggregates ARRAY<STRUCT<key STRING, value FLOAT64>>>> AS (
-(
-    WITH normalized AS (
-      SELECT
-        first_bucket,
-        last_bucket,
-        num_buckets,
-        latest_version,
-        metric,
-        metric_type,
-        key,
-        process,
-        agg_type,
-        udf_normalized_sum(aggregates) AS aggregates
-      FROM UNNEST(arrs))
+WITH filtered_data AS (
+  SELECT
+    sample_id,
+    client_id,
+    os,
+    app_version,
+    app_build_id,
+    channel,
+    first_bucket,
+    last_bucket,
+    num_buckets,
+    metric,
+    metric_type,
+    key,
+    process,
+    agg_type,
+    aggregates,
+    os = 'Windows'and channel = 'release' AS sampled
+  FROM
+    clients_histogram_aggregates_v1
+  CROSS JOIN UNNEST(histogram_aggregates)
+  WHERE submission_date = @submission_date
+    AND first_bucket IS NOT NULL),
+
+aggregated_histograms AS
```

Comment:
Does this just move the data cubing computation from probe counts into this query to prevent double counting? Is it possible to collapse the queries together now that the query is processing less data?

Reply:
Yes, the data cubing is moved so it is done per-client first to avoid double counting.

Unfortunately, it doesn't seem possible to collapse the queries. Initially I tried running with a 1% sample and it worked out. However, with the 10% sample, there are shuffle errors.

In fact, doing the data cubing per client for 10% seemed to give shuffle errors even without merging both bucket count + probe count queries. So I have made some changes that I will be pushing shortly to break up the processing into quarters like with histogram aggregates.

Reply:
Also worth noting that data cubing by client increased the amount of processing and therefore the cost by about $100 for the day I tested on.

Reply:
If you break up the query into smaller pieces, remember that the release windows clients are all taken from consecutive sample ids, so partitions will be uneven unless its adjusted e.g. `sample_id % x == 0`

And where does the extra 20TB of data come from?

Reply:
Here is what I tested:
The `clients_histogram_aggregates` table had ~15.5T of data based on a 10% sample of windows + release.
a) I ran `histogram_bucket_counts` as it appears now in this PR and that consistently gave a shuffle error.
b) I added a filter for `histogram_bucket_counts` between 0-24 inclusive sample_id  and this ran and processed 15T
c) I then separately ran just the first subquery (without the data cubing) and that processed 6T.

So even though both queries b) and c) should be reading the same amount of data from `clients_histogram_aggregates`, they had very different amounts of data processed. I could only speculate that this was due to internally the `filtered_data` subquery being run multiple times perhaps. Do you have any other intuitions for what this could be?

Based on this, I decided to have the `filtered_data` subquery as its own materialized table.
Now I ran the materialized table for partitioned on sample_id for *all*  sample_ids. This should process ~15.5T (the size of `clients_histogram_aggregates`) - It seemed to have processed 16.44T. I then ran the rest of `histogram_bucket_counts` and `histogram_probe_counts` as one query for each of the quarters of data and this processed a total of 22.46T (again larger than the 16.44T I would have expected, but less dramatic than 6T -> 15T. and maybe a similar issue where the table is being read multiple times with the data cubing).

Overall this cost 22.46T + 16.44T = 38.9T.
If we could just do the first read we wouldn't need that extra 22.46T to be processed.

To confirm that this was cheaper than not having the intermediate materialized table, I also ran `histogram_bucket_counts` with sample_ids between 50-74 inclusive, these would not have included samples. This was 11.7T.
So the cost without the materialized table would be 11.7T * 3 + 15T = 50.1T. So the materialized table saves ~11T.

That's a lot of text but if this happens to give you any money-saving ideas, let me know!

Also, good idea doing %x to keep it even.

Reply:
One more thing, I think the data cubing was a lot less costly before because by the time the data made it into `clients_histogram_bucket_counts` it was around 50GB! So even if that exploded with cubing, it wasn't too crazy.

Reply:
Based on the [execution details page](https://console.cloud.google.com/bigquery?project=glam-desktop-dev&j=bq:US:bquxjob_5991e116_171849d0c1f&page=queryresults) for the job, we can confirm that the filtered_data subquery is reading from the source table several times (3 times to be exact).

It might be useful to cross join twice on two smaller tables that only contain a null value for `os` and `app_build_id`. I'm not sure what the exact syntax would be off the top of my head, but something like

```sql
SELECT
  ...
FROM
   filtered_data
CROSS JOIN (SELECT null as os)
USING (os)
CROSS JOIN (SELECT null as app_build_id)
USING (app_build_id)
```

Reply:
I'm not 100% sure what your intention was with this example since `CROSS JOIN` doesn't seem to work with `USING` syntax. I tried your suggestion that you mentioned in slack with unioning the datasets before the group by [here](https://console.cloud.google.com/bigquery?project=glam-desktop-dev&j=bq:US:bquxjob_33f796c0_17198e1523a&page=queryresults). Unfortunately it seemed to have the same number of bytes processed as [here](https://console.cloud.google.com/bigquery?project=glam-desktop-dev&j=bq:US:bquxjob_7672ddad_17198ac1dd1&page=queryresults) where there are multiple group by clauses.

Reply:
Very cool that seems to have worked! Thanks for this idea!! And now I've updated the PR to get rid of `histogram_bucket_counts`. We can probably get rid of bucket counts for scalars too but this PR is getting hefty and I want to get it out the door

Reply:
Do you mind sharing the processed/slot time/shuffled/spilled stats (or links to the specific execution) for the baseline vs updated versions of the query when you get a chance? I'm curious how it turned out.

Reply:
Yea sure.

[Without static combos](https://console.cloud.google.com/bigquery?project=glam-desktop-dev&j=bq:US:bquxjob_7672ddad_17198ac1dd1&page=queryresults): 482.3gb processed.
[With static combos](https://console.cloud.google.com/bigquery?project=glam-desktop-dev&organizationId=442341870013&j=bq:US:bquxjob_aee9f96_171a816853d&page=queryresults): 120.6gb processed